### PR TITLE
Add haschrome=1 as query parameter for requests sent to authorize endpoint. 

### PR DESCRIPTION
**Pull request by weijjia**
#617

1. For mobile devices, there is already a physical back button, ADAL should send the query parameter haschrome=1 to hide the back button on AAD signin page thus developer will stop getting "access_denied" when it's actually user cancel the flow. 
2. Update tests.

### CODE REVIEW

**Comment by weijjia** [comment]
@heidijinxujia ^^^^

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java [hunk 1]
```diff
@@ -57,12 +57,25 @@
...
     private static final String TEST_RETURNED_EXCEPTION = "test-returned-exception";
 
     private static final String TEST_AUTHORITY = "https://login.windows.net/common";
+    @Override
+    protected void setUp() throws Exception {
+        super.setUp();
+        getContext().getCacheDir();
+        System.setProperty("dexmaker.dexcache", getContext().getCacheDir().getPath());
+        getContext().getCacheDir();
+        System.setProperty("dexmaker.dexcache", getContext().getCacheDir().getPath());
```
Comment (addev-ashish) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: why are you setting the same property twice?

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java [hunk 2]
```diff
@@ -162,215 +173,198 @@ public void testParseIdTokenNegativeInvalidEncodedTokens() {
...
                 "loginhint", null, null, null);
-        Object oauth = createOAuthInstance(request);
-        Method m = ReflectionUtils.getTestMethod(oauth, "getToken", String.class);
-
-        AuthenticationResult actual = (AuthenticationResult) m.invoke(oauth, "http://www.nocodeurl.com?state=YT1odHRwczovL2xvZ2luLndpbmRvd3MubmV0L2NvbW1vbiZyPWh0dHBzOi8vb2ZmaWNlYXBwcy5saXZlLmNvbQ");
-        assertNull(actual);
+        final Oauth2 oauth2 = createOAuthInstance(request);
+        final String uriWithNoCode = "http://www.nocodeurl.com?state=YT1odHRwczovL2xvZ2luLndpbmRvd3MubmV0L2NvbW1vbiZyPWh0dHBzOi8vb2ZmaWNlYXBwcy5saXZlLmNvbQ";
+        try {
+            final AuthenticationResult result = oauth2.getToken(uriWithNoCode);
```
Comment (addev-ashish) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: placeholder: will check on codeflow why so many empty lines here

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java [hunk 3]
```diff
@@ -162,215 +173,198 @@ public void testParseIdTokenNegativeInvalidEncodedTokens() {
...
+        final Oauth2 oauthWithoutLoginHint = createOAuthInstance(requestWithoutLogin);
 
-        actual = (String) m.invoke(oauthWithoutLoginHint);
+        final String actualCodeRequestNoLoginhint = oauthWithoutLoginHint.getCodeRequestUrl();
         assertTrue(
                 "Matching message",
-                actual.contains("http://www.something.com/oauth2/authorize?response_type=code&client_id=client+1234567890-%2B%3D%3B%21%23%24+++%26%27%28+%29*%2B%2C%2F%3A++%3B%3D%3F%40%5B%5D&resource=resource%2520urn%3A%21%23%24++++%26%27%28+%29*%2B%2C%2F%3A++%3B%3D%3F%40%5B%5D&redirect_uri=redirect+1234567890&state="));
-        assertFalse("Without loginhint", actual.contains("login_hint=loginhintForCode"));
+                actualCodeRequestNoLoginhint.contains(
+                        "http://www.something.com/oauth2/authorize?response_type=code&client_id=client+1234567890-%2B%3D%3B%21%23%24+++%26%27%28+%29*%2B%2C%2F%3A++%3B%3D%3F%40%5B%5D&resource=resource%2520urn%3A%21%23%24++++%26%27%28+%29*%2B%2C%2F%3A++%3B%3D%3F%40%5B%5D&redirect_uri=redirect+1234567890&state="));
```
Comment (addev-ashish) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: formatting seems to be off

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java [hunk 4]
```diff
@@ -162,215 +173,198 @@ public void testParseIdTokenNegativeInvalidEncodedTokens() {
...
-                PromptBehavior.Always, "&extra=1", null);
-
-        oauthExtraParam = createOAuthInstance(requestExtraParam);
-
-        actual = (String) m.invoke(oauthExtraParam);
-        assertTrue("Prompt", actual.contains("&prompt=login&extra=1"));
+                PromptBehavior.Always, "extra=1&haschrome=1", null);
+        final Oauth2 oAuthxtraQPContainHasChrome = createOAuthInstance(extraQPContainHasChrome);
+        final String actualCodeRequestQPHasChrome = oAuthxtraQPContainHasChrome.getCodeRequestUrl();
+        assertTrue("Prompt", actualCodeRequestQPHasChrome.contains("&prompt=login&extra=1&haschrome=1"));
```
Comment (addev-ashish) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: do you want to ignorecase in this check?

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java [hunk 5]
```diff
@@ -454,7 +446,7 @@ public void testRefreshTokenWebResponseDeviceChallengePositive()
                 + " Nonce=\"" + nonce + "\",  Version=\"1.0\", CertThumbprint=\"" + thumbPrint
                 + "\",  Context=\"" + context + "\"";
         String tokenPositiveResponse = "{\"access_token\":\"accessTokenHere\",\"token_type\":\"Bearer\",\"expires_in\":\"28799\",\"expires_on\":\"1368768616\",\"refresh_token\":\"refreshWithDeviceChallenge\",\"scope\":\"*\"}";
-        HashMap<String, List<String>> headers = getHeader(
+        Map<String, List<String>> headers = getHeader(
```
Comment (addev-ashish) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: final :)

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java [hunk 6]
```diff
@@ -498,158 +488,175 @@ public void testRefreshTokenWebResponseDeviceChallengeHeaderEmpty()
...
+        Util.prepareMockedUrlConnection(mockedConnection);
+        when(mockedConnection.getOutputStream()).thenReturn(mock(OutputStream.class));
+        when(mockedConnection.getInputStream()).thenReturn(
+                Util.createInputStream(jsonResponse));
+        when(mockedConnection.getResponseCode()).thenReturn(HttpURLConnection.HTTP_OK);
 
         // send call with mocks
         try {
-            AuthenticationResult result = (AuthenticationResult) m.invoke(oauth, mockResponse);
+            oauth2.refreshToken("fakeRefresToken");
```
Comment (addev-ashish) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: typo

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java (hunk 1 above)
Comment (weijjia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: my bad.

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java (hunk 2 above)
Comment (weijjia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: you can also check in editor :)

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java (hunk 3 above)
Comment (weijjia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: I should move up the Matching message, and move the contains also one line up. Will update

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java (hunk 4 above)
Comment (weijjia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: Nope, there is no need I think.

File: adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java (hunk 5 above)
Comment (weijjia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/OauthTests.java]: updated, and updated all the variables in the test.

**Comment by addev-ashish** [comment]
:shipit:

**Pull request merged**